jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterRepeatApply.java

```diff
@@ -18,99 +18,26 @@
 
 package org.apache.jena.sparql.engine.iterator;
 
-import java.util.NoSuchElementException;
-
-import org.apache.jena.atlas.lib.Lib;
-import org.apache.jena.atlas.logging.Log;
 import org.apache.jena.sparql.engine.ExecutionContext;
 import org.apache.jena.sparql.engine.QueryIterator;
 import org.apache.jena.sparql.engine.binding.Binding;
 
 /**
  * Repeatedly execute the subclass operation for each Binding in the input iterator.
  */
-public abstract class QueryIterRepeatApply extends QueryIter1 {
-    private int count = 0;
-    private QueryIterator currentStage;
-    private volatile boolean cancelRequested = false;
+public abstract class QueryIterRepeatApply extends QueryIterRepeatApplyBulk {
 
     public QueryIterRepeatApply(QueryIterator input, ExecutionContext context) {
         super(input, context);
-        this.currentStage = null;
-
-        if ( input == null ) {
-            Log.error(this, "[QueryIterRepeatApply] Repeated application to null input iterator");
-            return;
-        }
-    }
-
-    protected QueryIterator getCurrentStage() {
-        return currentStage;
     }
 
     protected abstract QueryIterator nextStage(Binding binding);
 
+    /** The caller guarantees that input has at least one more item */
     @Override
-    protected boolean hasNextBinding() {
-        if ( isFinished() )
-            return false;
-
-        for ( ;; ) {
-            if ( currentStage == null )
-                currentStage = makeNextStage();
-
-            if ( currentStage == null )
-                return false;
-
-            if ( cancelRequested )
-                // Pass on the cancelRequest to the active stage.
-                performRequestCancel(currentStage);
-
-            if ( currentStage.hasNext() )
-                return true;
-
-            // finish this step
-            currentStage.close();
-            currentStage = null;
-            // loop
-        }
-        // Unreachable
-    }
-
-    @Override
-    protected Binding moveToNextBinding() {
-        if ( !hasNextBinding() )
-            throw new NoSuchElementException(Lib.className(this) + ".next()/finished");
-        return currentStage.nextBinding();
-
-    }
-
-    private QueryIterator makeNextStage() {
-        count++;
-
-        if ( getInput() == null )
-            return null;
-
-        if ( !getInput().hasNext() ) {
-            getInput().close();
-            return null;
-        }
-
+    protected QueryIterator nextStage(QueryIterator input) {
```

Comment:
Please explain what is going on here. 

If a change is needed, then having a superclass is not necessary. Just two `nextStage` methods with different signatures.

Reply:
QueryIterApply**Bulk**'s `nextStage(QueryIterator)` method flat-maps the input `QueryIterator` to an output `QueryIterator` until the input iterator is consumed - i.e. implementations of nextStage can "per stage" consume an arbitrary amount > 0 from the input iterator.

QueryIterApply implements `QueryIterApplyBulk.nextStage(QueryIterator)` such that it delegates to its own usual `QueryIterApply.nextStage(Binding)` method.

Note that `QueryIterApplyBulk` is also the base class for `QueryIterServiceBulk`. The latter only operates on the `QueryIterator` and therefore does not need a `nextStage(Binding)` method. I find this separation cleaner than having both `nextStage` methods on the same class together with an override that causes the binding-level one to no longer be called.

The only drawback I see with this separation is that reflection code that relied on `QueryIterApply.class.getDeclaredMethods()` would break.

Reply:
> I find this separation cleaner

My concern is that this is making the bulk case "normal" when in fact it is special to your service work.